File: viaconnect-web/android/app/src/main/java/com/farmceutica/viaconnect/FormaVisionDepthPlugin.kt
// FormaVisionDepthPlugin.kt
// Task 14 (Prompt 210c) - ARCore Depth API booster plugin for Android.
//
// UNVERIFIED: This file requires:
//   1. A physical Android device with ARCore Depth API support (most flagship
//      Android 9+ phones with ToF sensors - check Google's list at
//      developers.google.com/ar/devices)
//   2. ARCore dependency added to android/app/build.gradle:
//        implementation 'com.google.ar:core:1.44.0'
//   3. Kotlin plugin in android/build.gradle:
//        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:<version>'
//      and in android/app/build.gradle:
//        apply plugin: 'kotlin-android'
//   4. Registration in MainActivity.java (see that file)
//
// Cannot be validated in the current web/Node environment.
//
// Graceful degradation: isDepthAvailable returns {available:false} on devices
// without ARCore or Depth API support. captureDepth rejects with a clear error.
// The JS layer (formaVisionDepth.ts probeDepthCapability) catches all errors.

package com.farmceutica.viaconnect

import android.graphics.ImageFormat
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.ar.core.ArCoreApk
import com.google.ar.core.Config
import com.google.ar.core.Session
import com.google.ar.core.TrackingState
import com.google.ar.core.exceptions.UnavailableArcoreNotInstalledException
import com.google.ar.core.exceptions.UnavailableDeviceNotCompatibleException
import com.google.ar.core.exceptions.UnavailableSdkTooOldException
import com.google.ar.core.exceptions.UnavailableUserDeclinedInstallationException
import java.nio.ByteOrder
import java.nio.ShortBuffer

@CapacitorPlugin(name = "FormaVisionDepth")
class FormaVisionDepthPlugin : Plugin() {

    // MARK: isDepthAvailable

    /**
     * Returns {available: boolean} indicating whether this device supports the
     * ARCore Depth API. Creates a one-shot ARCore Session just to check, then
     * closes it. Returns false on any exception.
     */
    @PluginMethod
    fun isDepthAvailable(call: PluginCall) {
        val ctx = context
        if (ctx == null) {
            val result = JSObject()
            result.put("available", false)
            call.resolve(result)
            return
        }
        val available = try {
            // Check ARCore availability first
            val availability = ArCoreApk.getInstance().checkAvailability(ctx)
            if (!availability.isSupported) {
                false
            } else {
                val session = Session(ctx)
                val supported = session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)
                session.close()
                supported
            }
        } catch (_: UnavailableArcoreNotInstalledException) {
            false
        } catch (_: UnavailableDeviceNotCompatibleException) {
            false
        } catch (_: UnavailableSdkTooOldException) {
            false
        } catch (_: UnavailableUserDeclinedInstallationException) {
            false
        } catch (_: Exception) {
            false
        }
        val result = JSObject()
        result.put("available", available)
        call.resolve(result)
    }

    // MARK: captureDepth

    /**
     * Starts an ARCore session with depth enabled, captures one frame with depth,
     * samples depth at the requested normalized Y levels, and returns a DepthFrame dict.
     *
     * Expected call options:
     *   levelNorms: array of doubles [0..1] - normalized Y positions to sample
     *
     * Returns a dict matching the DepthFrame interface in depthScale.ts.
     * Rejects when ARCore is unavailable, depth is not supported, or capture fails.
     *
     * IMPORTANT: This does a synchronous session.update() call. In a production
     * implementation, the session should run a few update() cycles on a dedicated
     * thread to let depth stabilize before sampling. Marked as a known limitation
     * for the Gary native-build review.
     */
    @PluginMethod
    fun captureDepth(call: PluginCall) {
        val levelNormsArr = call.getArray("levelNorms")
        if (levelNormsArr == null || levelNormsArr.length() == 0) {
            call.reject("levelNorms must be a non-empty array of normalized Y positions")
            return
        }

        val levelNorms = mutableListOf<Double>()
        for (i in 0 until levelNormsArr.length()) {
            levelNorms.add(levelNormsArr.getDouble(i))
        }

        val ctx = context
        if (ctx == null) {
            call.reject("Plugin context unavailable")
            return
        }

        // Run on a background thread to avoid blocking the main thread
        bridge.context.mainExecutor.execute {
            var session: Session? = null
            try {
                session = Session(ctx)

                if (!session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)) {
                    session.close()
                    call.reject("ARCore Depth API not supported on this device")
                    return@execute
                }

                val config = Config(session)
                config.depthMode = Config.DepthMode.AUTOMATIC
                // SEMANTIC_SCENE_UNDERSTANDING adds body segmentation but is optional
                session.configure(config)
                session.resume()

                // Attempt to capture a frame with valid depth.
                // Production recommendation: retry up to 10 frames (on a timed loop)
                // until the camera has tracked and depth is available.
                var depthImage: android.media.Image? = null
                var captureFrame: com.google.ar.core.Frame? = null
                var intrinsicsObj: JSObject? = null

                for (attempt in 0 until 5) {
                    val frame = session.update()
                    if (frame.camera.trackingState != TrackingState.TRACKING) {
                        Thread.sleep(100)
                        continue
                    }
                    captureFrame = frame

                    // Get camera intrinsics for the color (CPU) image
                    val imgIntrinsics = frame.camera.imageIntrinsics
                    val focal = imgIntrinsics.focalLength
                    val principal = imgIntrinsics.principalPoint
                    val dims = imgIntrinsics.imageDimensions
                    intrinsicsObj = JSObject().apply {
                        put("fx", focal[0].toDouble())
                        put("fy", focal[1].toDouble())
                        put("cx", principal[0].toDouble())
                        put("cy", principal[1].toDouble())
                        put("widthPx", dims[0].toDouble())
                        put("heightPx", dims[1].toDouble())
                    }

                    // Acquire depth image (DEPTH16 format)
                    try {
                        depthImage = frame.acquireDepthImage16Bits()
                        break
                    } catch (_: Exception) {
                        Thread.sleep(50)
                    }
                }

                val frame = captureFrame
                val depth = depthImage
                val intrinsics = intrinsicsObj

                if (depth == null || frame == null || intrinsics == null) {
                    session.pause()
                    session.close()
                    call.reject("Could not acquire a depth image after multiple attempts")
                    return@execute
                }

                val samples = sampleDepthImage(depth, levelNorms)
                depth.close()
                session.pause()
                session.close()

                val result = JSObject()
                result.put("samples", samples)
                result.put("intrinsics", intrinsics)
                result.put("capturedAtMs", System.currentTimeMillis().toDouble())
                call.resolve(result)

            } catch (e: Exception) {
                try {
                    session?.pause()
                    session?.close()
                } catch (_: Exception) { }
                call.reject("depth capture failed: ${e.message ?: "unknown error"}")
            }
        }
    }

    // MARK: Private helpers

    /**
     * Samples depth statistics at each requested normalized Y level.
     *
     * ARCore DEPTH16 format: each UInt16 pixel encodes:
     *   lower 13 bits = depth in millimeters (0 = invalid)
     *   upper 3 bits  = confidence (0 = low/invalid, 7 = highest confidence)
     *
     * This function accepts only pixels with confidence >= 1 and depth > 0.
     */
    private fun sampleDepthImage(depthImage: android.media.Image, levelNorms: List<Double>): JSArray {
        val width = depthImage.width
        val height = depthImage.height
        val plane = depthImage.planes[0]

        // DEPTH16 pixels are UInt16 in native byte order
        val rawBuffer = plane.buffer.order(ByteOrder.nativeOrder())
        val shortBuffer: ShortBuffer = rawBuffer.asShortBuffer()

        val samplesArr = JSArray()

        for (levelNorm in levelNorms) {
            val row = (levelNorm * (height - 1)).toInt().coerceIn(0, height - 1)
            val validDepths = mutableListOf<Double>()

            for (col in 0 until width) {
                val pixelIndex = row * width + col
                val rawVal = shortBuffer.get(pixelIndex).toInt() and 0xFFFF
                val depthMm = rawVal and 0x1FFF          // lower 13 bits = depth in mm
                val confidence = (rawVal shr 13) and 0x07  // upper 3 bits = confidence
                if (depthMm > 0 && confidence >= 1) {
                    validDepths.add(depthMm.toDouble() / 1000.0)  // mm to meters
                }
            }

            val sampleObj = JSObject()
            sampleObj.put("levelNorm", levelNorm)
            sampleObj.put("validPixelCount", validDepths.size)

            if (validDepths.size >= 5) {
                validDepths.sort()
                sampleObj.put("minDepthM", validDepths.first())
                sampleObj.put("maxDepthM", validDepths.last())
                sampleObj.put("medianDepthM", validDepths[validDepths.size / 2])
            } else {
                // Not enough valid pixels: return zeros (JS depthDerivedDepthCm returns null)
                sampleObj.put("minDepthM", 0.0)
                sampleObj.put("maxDepthM", 0.0)
                sampleObj.put("medianDepthM", 0.0)
            }

            samplesArr.put(sampleObj)
        }

        return samplesArr
    }
}

File: viaconnect-web/ios/App/App/FormaVisionDepthPlugin.swift
// FormaVisionDepthPlugin.swift
// Task 14 (Prompt 210c) - ARKit LiDAR depth booster plugin.
//
// UNVERIFIED: This file requires a physical iOS device with LiDAR (iPhone 12 Pro+
// or any iPad Pro with LiDAR) and a native Xcode build to verify. It cannot be
// validated in the current web/Node environment.
//
// Registration: see FormaVisionDepthPluginBridge.m for the CAP_PLUGIN macro.
// Both files must be added to the Xcode target for the plugin to function.
//
// Requirements:
//   - Xcode 14+ (for ARKit scene depth support in simulator stub)
//   - iOS 14+ deployment target (ARKit sceneDepth requires iOS 14)
//   - Real device with LiDAR for actual depth (simulator returns unavailable)
//   - Info.plist: NSCameraUsageDescription (already present) + NSMotionUsageDescription
//   - No additional Podfile entries needed (ARKit is a system framework)
//
// Graceful degradation: isDepthAvailable returns {available:false} on all devices
// without LiDAR. captureDepth rejects (JS catch block -> null) on non-LiDAR devices.
// The JS layer (formaVisionDepth.ts) never exposes depth to the pipeline on failure.

import Foundation
import Capacitor
import ARKit

// MARK: - Plugin

@objc(FormaVisionDepthPlugin)
public class FormaVisionDepthPlugin: CAPPlugin {

    // Retain the AR delegate while a capture is in flight.
    // One capture at a time; concurrent calls are not supported (first wins).
    private var activeCaptureDelegate: FormaVisionARDelegate?

    // MARK: isDepthAvailable

    /// Returns { available: boolean } indicating whether this device supports ARKit
    /// scene depth (LiDAR). ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth)
    /// returns false on all non-LiDAR devices and in the iOS simulator.
    @objc func isDepthAvailable(_ call: CAPPluginCall) {
        let supported = ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth)
        call.resolve(["available": supported])
    }

    // MARK: captureDepth

    /// Starts an ARSession with scene depth semantics, waits for the first valid depth frame,
    /// samples depth at the requested normalized Y levels, and returns a DepthFrame dict.
    ///
    /// Expected call options:
    ///   levelNorms: [Double]  - normalized Y positions [0..1] to sample
    ///
    /// Returns a dict matching the DepthFrame interface in depthScale.ts:
    ///   {
    ///     samples: [{ levelNorm, minDepthM, maxDepthM, medianDepthM, validPixelCount }],
    ///     intrinsics: { fx, fy, cx, cy, widthPx, heightPx },
    ///     capturedAtMs: Double
    ///   }
    ///
    /// Rejects when:
    ///   - device does not support scene depth (LiDAR absent)
    ///   - levelNorms is missing or empty
    ///   - ARSession fails to start or produce a valid depth frame within timeout
    @objc func captureDepth(_ call: CAPPluginCall) {
        guard ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth) else {
            call.reject("LiDAR scene depth is not supported on this device")
            return
        }

        guard let rawLevelNorms = call.getArray("levelNorms"),
              rawLevelNorms.count > 0 else {
            call.reject("levelNorms must be a non-empty array of normalized Y positions [0..1]")
            return
        }

        // Parse levelNorms as Doubles
        var levelNorms: [Double] = []
        for i in 0..<rawLevelNorms.count {
            if let v = rawLevelNorms[i] as? Double {
                levelNorms.append(v)
            } else if let v = rawLevelNorms[i] as? NSNumber {
                levelNorms.append(v.doubleValue)
            }
        }
        guard !levelNorms.isEmpty else {
            call.reject("levelNorms contained no valid Double values")
            return
        }

        // Retain the delegate until capture completes (success or error)
        let delegate = FormaVisionARDelegate(levelNorms: levelNorms) { [weak self] result in
            self?.activeCaptureDelegate = nil
            call.resolve(result)
        } onError: { [weak self] message in
            self?.activeCaptureDelegate = nil
            call.reject(message)
        }
        activeCaptureDelegate = delegate

        // ARSession must be started on the main thread
        DispatchQueue.main.async {
            let session = ARSession()
            session.delegate = delegate
            let config = ARWorldTrackingConfiguration()
            config.frameSemantics = [.sceneDepth]
            // Run with resetTracking to force a fresh session for each capture
            session.run(config, options: [.resetTracking, .removeExistingAnchors])
            // Store the session in the delegate so it can be paused on completion
            delegate.session = session
        }
    }
}

// MARK: - AR Session Delegate

/// Waits for the first valid ARFrame with a sceneDepth map, samples depth at
/// the requested body levels, and invokes either onResult or onError exactly once.
private final class FormaVisionARDelegate: NSObject, ARSessionDelegate {
    private let levelNorms: [Double]
    private let onResult: ([String: Any]) -> Void
    private let onError: (String) -> Void
    private var resolved = false

    // Assigned after session.run() is called (to allow session.pause() on completion)
    weak var session: ARSession?

    init(
        levelNorms: [Double],
        onResult: @escaping ([String: Any]) -> Void,
        onError: @escaping (String) -> Void
    ) {
        self.levelNorms = levelNorms
        self.onResult = onResult
        self.onError = onError
    }

    // Called on ARKit's internal queue for each new frame
    func session(_ arSession: ARSession, didUpdate frame: ARFrame) {
        guard !resolved else { return }
        // Wait for a frame that has a valid depth map
        guard let sceneDepth = frame.sceneDepth,
              let depthMap = Optional(sceneDepth.depthMap) else {
            return
        }

        resolved = true
        arSession.pause()

        processFrame(frame: frame, depthMap: depthMap)
    }

    func session(_ arSession: ARSession, didFailWithError error: Error) {
        guard !resolved else { return }
        resolved = true
        arSession.pause()
        onError("ARSession failed: \(error.localizedDescription)")
    }

    // MARK: Private processing

    private func processFrame(frame: ARFrame, depthMap: CVPixelBuffer) {
        // Camera intrinsics (simd_float3x3, column-major):
        //   [0][0] = fx, [1][1] = fy, [2][0] = cx, [2][1] = cy
        let m = frame.camera.intrinsics
        let resolution = frame.camera.imageResolution
        let intrinsicsDict: [String: Any] = [
            "fx": Double(m[0][0]),
            "fy": Double(m[1][1]),
            "cx": Double(m[2][0]),
            "cy": Double(m[2][1]),
            "widthPx": Double(resolution.width),
            "heightPx": Double(resolution.height),
        ]

        // Depth map dimensions (may differ from color image - typically smaller)
        let depthWidth = CVPixelBufferGetWidth(depthMap)
        let depthHeight = CVPixelBufferGetHeight(depthMap)

        CVPixelBufferLockBaseAddress(depthMap, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(depthMap, .readOnly) }

        guard let baseAddr = CVPixelBufferGetBaseAddress(depthMap) else {
            onError("depth map base address unavailable")
            return
        }

        // ARKit depth maps are kCVPixelFormatType_DepthFloat32 - each pixel is a Float32
        // metric depth value in meters (0 = invalid/unknown)
        let floatPtr = baseAddr.assumingMemoryBound(to: Float32.self)

        var samplesArr: [[String: Any]] = []

        for levelNorm in levelNorms {
            // Map normalized Y in color-image space to the depth map row.
            // Because levelNorm spans [0..1], applying it to depthHeight gives the
            // correct proportional row regardless of depth/color resolution ratio.
            let row = Int(levelNorm * Double(depthHeight - 1)).clamped(to: 0...(depthHeight - 1))
            var validDepths: [Double] = []

            for col in 0..<depthWidth {
                let idx = row * depthWidth + col
                let d = Double(floatPtr[idx])
                // ARKit uses 0.0 for invalid/unknown pixels
                if d.isFinite && d > 0.001 {
                    validDepths.append(d)
                }
            }

            var sampleDict: [String: Any] = [
                "levelNorm": levelNorm,
                "validPixelCount": validDepths.count,
            ]

            if validDepths.count >= 5 {
                validDepths.sort()
                sampleDict["minDepthM"] = validDepths.first!
                sampleDict["maxDepthM"] = validDepths.last!
                sampleDict["medianDepthM"] = validDepths[validDepths.count / 2]
            } else {
                // Not enough valid pixels: return zeros (JS depthDerivedDepthCm returns null)
                sampleDict["minDepthM"] = 0.0
                sampleDict["maxDepthM"] = 0.0
                sampleDict["medianDepthM"] = 0.0
            }

            samplesArr.append(sampleDict)
        }

        let result: [String: Any] = [
            "samples": samplesArr,
            "intrinsics": intrinsicsDict,
            "capturedAtMs": Date().timeIntervalSince1970 * 1000.0,
        ]
        onResult(result)
    }
}

// MARK: - Comparable range clamp helper

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        return max(range.lowerBound, min(self, range.upperBound))
    }
}
